src/main/java/com/faforever/api/security/FafAuthenticationConverter.java

```diff
@@ -1,29 +1,35 @@
 package com.faforever.api.security;
 
-import java.util.List;
-import java.util.Map;
-
 import org.springframework.core.convert.converter.Converter;
 import org.springframework.security.authentication.AbstractAuthenticationToken;
 import org.springframework.security.oauth2.jwt.Jwt;
 
+import java.util.List;
+import java.util.Map;
+
 /**
  * Jwt converter that reads scopes + custom FAF roles from the token extension.
  */
 public class FafAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {
 
   @Override
   public AbstractAuthenticationToken convert(Jwt source) {
-    int userId = extractUserId(source);
-    String username = extractUsername(source);
     List<FafScope> scopes = extractScopes(source);
     List<FafRole> roles = extractRoles(source);
 
-    return new FafAuthenticationToken(userId, username, scopes, roles);
+    String subject = extractSubject(source);
+
+    try {
+      int userId = Integer.parseInt(subject);
+      String username = extractUsername(source);
+      return new FafUserAuthenticationToken(userId, username, scopes, roles);
+    } catch (NumberFormatException e) {
+      return new FafServiceAuthenticationToken(subject, scopes);
+    }
```

Comment:
Do we want to perform the type discrimination off a scope or some value in ext rather than just assuming if there is an exception that it is a service?

Maybe even just checking if the subject is a numeric string might read better and convey the intention rather than now where it looks like the Service is just a fallback instead of a distinct choice.

Reply:
No I think that check is misleading. We (user-service + hydra) are the only authority to hand out token. If we mess up I'd rather see things explode than having a fallback state where only some things work but we don't really know what is going on.

Reply:
I think we are agreeing that we shouldn't just fallback to assuming it is a ServiceAuthenticationToken, just have to agree on how to check it and throw an exception if the discriminator fails

Reply:
No was the answer to "Do we want to perform ..."  :D So yes we are agreeing

Reply:
Then wouldn't the answer be yes we do want to perform explicit type discrimination XD

Reply:
well this was my best effort with token I got from hydra with client credentials.

Reply:
No this converter class is what produces the authentication from the jwt token that we have been given. So happens as the first step in the process.

We have no information about the token besides what is in the token.

So we need so kind of discriminator to tell which kind of authentication the token is representing which I suggest we put into the ext claims when we create the token in hydra.

Reply:
could we add a service claim to the ext? Or is that just the webhook you were talking about.

Reply:
Currently we can'T do anything to m2m tokens.